### PR TITLE
Two Person Mecha for Robotics: The Only Real Solution To An Imminent Geometer Summoning 💪 

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

Coop mecha is back, post vehicle refactor. Very minimal code was actually needed to do this, which is super nice.

- [x] Add special function for pilot: skyfall
- [x] Add special function for gunner: ivanov strike
- [x] make sure driver cannot change gunner's equipment
- [x] doc it... aw nuts.
- [x] swap seat action

## Why It's Good For The Game

Reread the title again if you don't think this shit is tight as hell. I mean shit man, what else am I here to say. Robotics starts with two-three people and it just feels so right to team up with others for a big advantage. There are countless ways this can go wrong, so it balances out.

## Changelog
:cl:
add: Robotics Two Person Mecha!
/:cl:

### CODE REVIEW

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	wreckage = /obj/structure/mecha_wreckage/savannah_ivanov
+	internal_damage_threshold = 25
+	max_occupants = 2
+
+/obj/vehicle/sealed/mecha/combat/savannah_ivanov/get_mecha_occupancy_state()
+	var/driver_present = length(driver_amount()) > 0
+	var/gunner_present = length(return_controllers_with_flag(VEHICLE_CONTROL_EQUIPMENT)) > 0
+	return "[base_icon_state]_[driver_present]_[gunner_present]"
+
+/obj/vehicle/sealed/mecha/combat/savannah_ivanov/auto_assign_occupant_flags(mob/M)
```
Comment (Qustinnus) [inline code/modules/vehicles/mecha/combat/savannah_ivanov.dm:31]: Can you add seat switching?

Comment (tralezab) [reply code/modules/vehicles/mecha/combat/savannah_ivanov.dm:31]: Sure!

File: code/modules/vehicles/mecha/mecha_actions.dm [hunk 2]
```diff
@@ -216,13 +224,13 @@
...
 	switch(chassis.damtype)
 		if("tox")
-			new_damtype = "brute"
+			new_damtype = BRUTE
 			to_chat(owner, "[icon2html(chassis, owner)]<span class='notice'>Your exosuit's hands form into fists.</span>")
 		if("brute")
-			new_damtype = "fire"
+			new_damtype = BURN
 			to_chat(owner, "[icon2html(chassis, owner)]<span class='notice'>A torch tip extends from your exosuit's hand, glowing red.</span>")
 		if("fire")
```
Comment (Fikou) [inline code/modules/vehicles/mecha/mecha_actions.dm:232]: you set them to defines but you dont check for those defines here

Comment (tralezab) [reply code/modules/vehicles/mecha/mecha_actions.dm:232]: Good point

**Comment by Greniza** [comment]
Could we get images of the sprites? I wanna see, but I don't wanna pull the PR.

**Comment by Twaticus** [comment]
image

File: code/modules/vehicles/mecha/_mecha.dm [hunk 3]
```diff
@@ -515,13 +521,13 @@
...
 		return
-	if(completely_disabled || is_currently_ejecting)
+	if(completely_disabled || is_currently_ejecting || (mecha_flags & CANNOT_INTERACT))
 		return
 	var/list/modifiers = params2list(params)
 	if(isAI(user) == !LAZYACCESS(modifiers, MIDDLE_CLICK))//BASICALLY if a human uses MMB, or an AI doesn't, then do nothing.
 		return
 	if(phasing)
-		to_chat(occupants, "[icon2html(src, occupants)]<span class='warning'>Unable to interact with objects while phasing.</span>")
+		to_chat(occupants, "[icon2html(src, occupants)]<span class='warning'>Unable to interact with objects while [phasing].</span>")
```
Comment (InsaneRed) [inline code/modules/vehicles/mecha/_mecha.dm:530]: maybe use ballon here?

Comment (InsaneRed) [reply code/modules/vehicles/mecha/_mecha.dm:530]: or ideally at every other to_chat

Comment (tralezab) [reply code/modules/vehicles/mecha/_mecha.dm:530]: will do, good suggestion

**Comment by Zytolg** [comment]
Now connect their brains to the robot.

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+		tiled_turf.break_tile()
+	var/turf/landed_on = get_turf(src)
+	if(isclosedturf(landed_on))
+		landed_on.ScrapeAway()
+	for(var/mob/living/victim in landed_on)
+		to_chat(victim, "<span class='userdanger'>[src] lands on you from above!</span>")
+		if(victim.stat != CONSCIOUS)
+			victim.gib(FALSE, FALSE, FALSE)
+		else
+			victim.adjustBruteLoss(80)
```
Comment (SmArtKar) [inline code/modules/vehicles/mecha/combat/savannah_ivanov.dm:161]: Make it damage not only living mobs, but all the objects and vehicles, else it's useless in mecha vs mecha combat. Maybe extend the damage from just one tile to 3x3 zone of landing

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+/obj/vehicle/sealed/mecha/combat/savannah_ivanov/proc/start_missile_targeting(mob/gunner, silent = TRUE)
+	if(!silent)
+		to_chat(gunner, "<span class='warning'>Ivanov Strike targeting process booted.</br>\
+		Your next click will fire the missile (provided the mech is facing the right direction).</span>")
+	aiming_ivanov = TRUE
+	RegisterSignal(src, COMSIG_MECHA_MELEE_CLICK, .proc/on_melee_click)
+	RegisterSignal(src, COMSIG_MECHA_EQUIPMENT_CLICK, .proc/on_equipment_click)
+	gunner.client.mouse_override_icon = 'icons/effects/mouse_pointers/supplypod_down_target.dmi'
+	gunner.update_mouse_pointer()
+	gunner.overlay_fullscreen("ivanov", /atom/movable/screen/fullscreen/ivanov_display, 1)
```
Comment (SmArtKar) [inline code/modules/vehicles/mecha/combat/savannah_ivanov.dm:178]: Overlay won't fully cover screen for those with widescreen preference toggled

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+/obj/vehicle/sealed/mecha/combat/savannah_ivanov/proc/drop_missile(mob/gunner, turf/target_turf)
+	end_missile_targeting(gunner)
+	SEND_SOUND(gunner, 'sound/machines/triple_beep.ogg')
+	COOLDOWN_START(src, strike_cooldown, strike_cooldown_time)
+	podspawn(list(
+		"target" = target_turf,
+		"style" = STYLE_MISSILE,
+		"effectMissile" = TRUE,
+		"explosionSize" = list(0,0,3,3)
+	))
```
Comment (SmArtKar) [inline code/modules/vehicles/mecha/combat/savannah_ivanov.dm:213]: Just a suggestion, maybe you can make the missiles weaker but allow to drop 2-3 of them?

**Comment by moocowswag** [comment]
Can a solo roboticist put a posibrain or mmi in one (or both) of the seats?

**Comment by tralezab** [comment]
@moocowswag They can, but I actually remember saying I would disable that for the old pr. It's kinda cheesy and removes an important balancing angle of the mech if you can just print a loyal co-pilot

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm (hunk 4 above)
Comment (tralezab) [reply code/modules/vehicles/mecha/combat/savannah_ivanov.dm:161]: Alright, I'll buff it

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm (hunk 6 above)
Comment (tralezab) [reply code/modules/vehicles/mecha/combat/savannah_ivanov.dm:213]: Gonna sound like a baby here but if you want it, you should pr it to my branch

**Comment by tralezab** [comment]
savannah mech is now all documented!

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm (hunk 5 above)
Comment (tralezab) [reply code/modules/vehicles/mecha/combat/savannah_ivanov.dm:178]: this is... either an issue with all fullscreens, or none of them. either way, not MY problem >:]

Comment (tralezab) [reply code/modules/vehicles/mecha/combat/savannah_ivanov.dm:241]: I checked, it ends up covering widescreen! 🎉

**Comment by GunHog** [comment]
If the gunner tries to eject, will he be interrupted by the pilot moving the mech?

How does this mech handle AIs? Can the AI swap "seats" as a human can? If a Malf AI hijacks this mech, will one or both occupants be ejected?

**Review by Twaticus — APPROVED** [review]
fortnite mech

**Comment by tralezab** [comment]
why 😢

**Comment by dondonpow** [comment]
you gotta be faster, coder man

**Comment by tralezab** [comment]
it's done!

File: code/__DEFINES/vehicles.dm [hunk 7]
```diff
@@ -1,8 +1,18 @@
...
+#define VEHICLE_CONTROL_KIDNAPPED (1<<1)
+///melee attacks/shoves a vehicle may have
+#define VEHICLE_CONTROL_MELEE (1<<2)
+///using equipment/weapons on the vehicle
+#define VEHICLE_CONTROL_EQUIPMENT (1<<3)
+///changing around settings and the like.
+#define VEHICLE_CONTROL_SETTINGS (1<<4)
+
+///ez define for giving a single pilot mech all the flags it needs.
+#define FULL_MECHA_CONTROL VEHICLE_CONTROL_DRIVE|VEHICLE_CONTROL_SETTINGS|VEHICLE_CONTROL_SETTINGS|VEHICLE_CONTROL_MELEE|VEHICLE_CONTROL_EQUIPMENT
```
Comment (Mothblocks) [inline code/__DEFINES/vehicles.dm:15]: You could just use ALL, but whatever.

File: code/modules/research/techweb/all_nodes.dm [hunk 8]
```diff
@@ -760,6 +760,26 @@
 	"phazon_peri", "phazon_targ", "phazon_armor")
 	research_costs = list(TECHWEB_POINT_TYPE_GENERIC = 2500)
 
+/datum/techweb_node/savannah_ivanov
+	id = "mecha_savannah_ivanov"
+	display_name = "EXOSUIT: Savannah-Ivanov"
+	description = "Savannah-Ivanov exosuit designs"
+	prereq_ids = list("adv_mecha", "weaponry", "exp_tools")
+	design_ids = list(
+		"savannah_ivanov_chassis",
```
Comment (Mothblocks) [inline code/modules/research/techweb/all_nodes.dm:769]: The change for this is about to be merged, make all these ABC order with a trailing comma for the last one.

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm [hunk 9]
```diff
@@ -0,0 +1,385 @@
+
+
+///how much time between charge_level going up by 1
```
Comment (Mothblocks) [inline code/modules/vehicles/mecha/combat/savannah_ivanov.dm:3]:
```suggestion
///how much time between charge_level going up by 1
```

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+#define SKYFALL_SINGLE_CHARGE_TIME 2 SECONDS
```
Comment (Mothblocks) [inline code/modules/vehicles/mecha/combat/savannah_ivanov.dm:4]: #undef all of these at the end.

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+		end_missile_targeting(getting_out)
+	. = ..()
+
+/**
+ * ## begin_skyfall_charge
+ *
+ * Proc called by the mecha's ability that starts the skyfall loop
+ */
+/obj/vehicle/sealed/mecha/combat/savannah_ivanov/proc/begin_skyfall_charge(mob/pilot)
+	to_chat(pilot, "[icon2html(src, pilot)]<span class='notice'>Charging Skyfall. Moving will interrupt the charge.</span>")
```
Comment (Mothblocks) [inline code/modules/vehicles/mecha/combat/savannah_ivanov.dm:85]: Want to change this to balloon alerts?

File: code/modules/vehicles/mecha/combat/savannah_ivanov.dm [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+
+/datum/action/vehicle/sealed/mecha/ivanov_strike/Trigger()
+	if(!owner || !chassis || !(owner in chassis.occupants))
+		return
+	var/obj/vehicle/sealed/mecha/combat/savannah_ivanov/ivanov_mecha = chassis
+	if(!COOLDOWN_FINISHED(ivanov_mecha, strike_cooldown))
+		var/timeleft = COOLDOWN_TIMELEFT(ivanov_mecha, strike_cooldown)
+		to_chat(owner, "<span class='warning'>You need to wait [DisplayTimeText(timeleft, 1)] before firing another Ivanov Strike.</span>")
+		return
+	ivanov_mecha.aiming_ivanov ? ivanov_mecha.end_missile_targeting(owner) : ivanov_mecha.start_missile_targeting(owner)
```
Comment (Mothblocks) [inline code/modules/vehicles/mecha/combat/savannah_ivanov.dm:344]: Just make this an if-statement, using ternaries like this looks wrong.

File: code/modules/vehicles/mecha/mecha_actions.dm [hunk 13]
```diff
@@ -236,7 +244,37 @@
...
 	button_icon_state = "mech_phasing_[chassis.phasing ? "on" : "off"]"
 	chassis.balloon_alert(owner, "[chassis.phasing ? "enabled" : "disabled"] phasing")
 	UpdateButtonIcon()
+
+///swap seats, for two person mecha
+/datum/action/vehicle/sealed/mecha/swap_seat
+	name = "Switch Seats"
+	button_icon_state = "mech_seat_swap"
+
+/datum/action/vehicle/sealed/mecha/swap_seat/Trigger()
```
Comment (Mothblocks) [inline code/modules/vehicles/mecha/mecha_actions.dm:257]: Change everything in here to balloon alerts.

File: code/modules/vehicles/mecha/mecha_construction_paths.dm [hunk 14]
```diff
@@ -1202,6 +1202,148 @@
...
+		if(21)
+			if(diff==FORWARD)
+				user.visible_message("<span class='notice'>[user] secures Savannah-Ivanov Armor Plates.</span>", "<span class='notice'>You secure Savannah-Ivanov Armor Plates.</span>")
+			else
+				user.visible_message("<span class='notice'>[user] pries Savannah-Ivanov Armor Plates from [parent].</span>", "<span class='notice'>You pry Savannah-Ivanov Armor Plates from [parent].</span>")
+		if(22)
+			if(diff==FORWARD)
+				user.visible_message("<span class='notice'>[user] welds Savannah-Ivanov Armor Plates to [parent].</span>", "<span class='notice'>You weld Savannah-Ivanov Armor Plates to [parent].</span>")
+			else
+				user.visible_message("<span class='notice'>[user] unfastens Savannah-Ivanov Armor Plates.</span>", "<span class='notice'>You unfasten Savannah-Ivanov Armor Plates.</span>")
```
Comment (Mothblocks) [inline code/modules/vehicles/mecha/mecha_construction_paths.dm:1344]: people dont actually code like this. we dont have code that looks like tihs. I dont believe it

**Review by Mothblocks — CHANGES_REQUESTED** [review]
code loooks fine

File: code/modules/vehicles/mecha/mecha_construction_paths.dm (hunk 14 above)
Comment (tralezab) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:1344]: lmao

File: code/__DEFINES/vehicles.dm (hunk 7 above)
Comment (tralezab) [reply code/__DEFINES/vehicles.dm:15]: eh sure

**Comment by Mothblocks** [comment]
arm gets to make this change because he knows what cho'gall is (heroes of the storm)

**Comment by NullFag** [comment]
> @moocowswag They can, but I actually remember saying I would disable that for the old pr. It's kinda cheesy and removes an important balancing angle of the mech if you can just print a loyal co-pilot

mmi's/posibrains aren't lawed

**Comment by Akride** [comment]
Hey ! I fucking love this mech it's going to be so so cool when it hits fulp ! One question i'm asking is : when using skyfall or ivannov strike, the mech straight up loses its internal components or its equipment. Could you maybe explain that mechanic ? Not sure if it's a bug or if it's intended. For example, after using skyfall, I will be unable to move, for the very good reason that the mech doesn't have a battery, sensor, and capacitator anymore.

**Comment by TheSilverNuke** [comment]
I have a similar issue. This happens every time and it actually killed the posi I inserted as the pilot once.

**Pull request merged**